### PR TITLE
feat: add gofumpt

### LINKED ISSUE #604: Formatting with gofumpt
There have been discussions previously about potentially formatting (and enforcing formatting) with [gofumpt](https://github.com/mvdan/gofumpt) (see tendermint/tendermint#9738, tendermint/tendermint#9208, tendermint/tendermint#8478).

The main argument against this was that doing a big-bang reformat of the whole codebase significantly hampers our ability to do code archeology, which, with a codebase like this, is incredibly important for our team to be able to make meaningful progress.

A compromise here would be to start to enforce this locally - i.e. on each of our own machines when saving our code, that only the files each of us touch in a specific PR get formatted with gofumpt. As we progressively refactor the system, I'm fairly confident we'll end up touching most of the codebase, simultaneously progressively formatting the code.

At some point, once the diff between formatting and not formatting the codebase with gofumpt drops below a certain level (say a few hundred LoC), we can simply format the remainder of the code in a single PR and implement a formatting enforcement check as part of CI (at which point we can close this issue).

### PR DESCRIPTION
**Pull request by melekes**
Our codebase is already sufficiently formatted, so we can enable **`gofumpt` by default**.

This PR also merges `lint-typo`, `lint-fix-typo`, `format`, `lint-format`, and `lint` Makefile
targets into a single `lint` target to ease the software development process. `codespell` CI job is removed in favor of `lint` job.

Closes #604

### CODE REVIEW

File: scripts/gofumpt.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+# Format staged Go files with `gofumpt`
+
+# Redirect output to stderr.
+exec 1>&2
```
Comment (mzabaluev) [inline scripts/gofumpt.sh:5]: Why not let the operator redirect streams the way they see fit? Especially as the script outputs to `>&2` as if it makes a difference after this.

**Review by mzabaluev — COMMENTED** [review]
Good, let's establish more order.

**Comment by faddat** [comment]
gofumpt is enforced by .golangci.yml currently

File: scripts/gofumpt.sh (hunk 1 above)
Comment (melekes) [reply scripts/gofumpt.sh]: Fixed in https://github.com/cometbft/cometbft/pull/2049/commits/0a73d731aaa9dfba29687dc28eaca1f64f95489b

**Comment by melekes** [comment]
> gofumpt is enforced by .golangci.yml currently

I find it interesting that the `gofumpt` linter in `golangci` does not cover all the files. I.e. when I run `gofumpt` separately, it will find and format more files than the linter in `golangci`. Hence, it's added to `Makefile` targets in this PR.

**Comment by melekes** [comment]
NOTE: `golangci-lint` CI job was renamed to `lint`

File: Makefile [hunk 2]
```diff
@@ -240,41 +242,28 @@ clean_certs:
...
-#	find . -name '*.go' -type f -not -path "*.git*"  -not -name '*.pb.go' -not -name '*pb_test.go' | xargs goimports -w -local github.com/cometbft/cometbft
-#.PHONY: format
-
-#? lint: Run latest golangci-lint linter
+#? lint: Lint, format and fix typos
 lint:
-	@echo "--> Running linter"
-	@go run github.com/golangci/golangci-lint/cmd/golangci-lint@latest run
+	@echo "--> Linting"
+	@go run github.com/golangci/golangci-lint/cmd/golangci-lint@latest run --fix
```
Comment (cason) [inline Makefile:248]: Are you sure you want to automatically fix here? I think I liked it fails when there are problems.

File: Makefile [hunk 3]
```diff
@@ -240,41 +242,28 @@ clean_certs:
...
+	@echo "--> Formatting"
+	@go run mvdan.cc/gofumpt@latest -l -w ./..
```
Comment (cason) [inline Makefile:250]: Not sure if I understood why `./..`.

Comment (cason) [reply Makefile:250]: Moreover, for some reason it reports here files inside `.git` dir.

**Review by cason — APPROVED** [review]
👍

File: Makefile (hunk 2 above)
Comment (melekes) [reply Makefile:248]: > Are you sure you want to automatically fix here?

Why would you not prefer the linter to fix the problem for you? `golangci-lint` will fail if the linter doesn't provide `AutoFix` https://golangci-lint.run/usage/linters/, which is largely still the case for many linters.

File: Makefile (hunk 3 above)
Comment (melekes) [reply Makefile]: fixed https://github.com/cometbft/cometbft/pull/2049/commits/fe80aad63d0095c95257399ac22441d7dabd0fe6

File: Makefile (hunk 2 above)
Comment (greg-szabo) [reply Makefile:248]: I agree that fixing is the right thing to do. If it causes issues down the line, we'll revert to no-fixing.

**Review by greg-szabo — APPROVED** [review]
I have more ideas that I would like to add/improve in a subsequent PR. Until then, this reduces chaos, it's a good thing.

**Comment by melekes** [comment]
FYI, I don't have a strong opinion about merging 3 workflows into 1 (golangci-lint, gofumpt, codespell). I can break them apart if needed.

**Pull request merged**